### PR TITLE
Polyfill the incremental generator ForAttributeWithMetadataName from roslyn (for JsonGenerator).

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/runtime/pull/70911

### CODE REVIEW

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 1]
```diff
@@ -550,38 +551,6 @@ private static bool TryGetClassDeclarationList(INamedTypeSymbol typeSymbol, [Not
...
-                        {
-                            continue;
-                        }
-
-                        INamedTypeSymbol attributeContainingTypeSymbol = attributeSymbol.ContainingType;
-                        string fullName = attributeContainingTypeSymbol.ToDisplayString();
-
-                        if (fullName == JsonSerializableAttributeFullName)
-                        {
-                            return classDeclarationSyntax;
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:576]: note: Teh only check we did was that hte fully qualified name is what we want.  that's exactly what roslyn now does, so we need no additional semantic checks once we move to the roslyn api.

File: src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    /// </summary>
+    internal ref partial struct ValueListBuilder<T>
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public T Pop()
+        {
+            _pos--;
+            return _span[_pos];
+        }
+    }
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs:19]: same comment as: https://github.com/dotnet/runtime/pull/71651/files#r914021221

**Comment by joperezr** [comment]
cc @eiriktsarpalis @layomia @krwq @ericstj  PTAL. For context, @CyrusNajmabadi is helping us add a much more performant way for source generators that depend on Attributes to run, and with these PRs he is onboarding generators one-by-one. Since you are the owners of the Json Source generator, do you mind taking a look at the changes and make sure they makes sense?

File: src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// These public methods are required by RegexWriter.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs:9]: Common seems specific to a different source generator.

**Review by eiriktsarpalis — APPROVED** [review]
LGTM other than a couple of comments.

File: src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs:9]: this file will go away once https://github.com/dotnet/runtime/pull/71652 is merged.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs:9]: SGTM

File: src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/libraries/System.Text.Json/src/System/Collections/Generic/ValueListBuilder.Pop.cs:19]: It def can, and i've made that change in https://github.com/dotnet/runtime/pull/71652.  I'll port that over to this one.

**Pull request merged**